### PR TITLE
#812 store export definitions for reuse

### PR DESCRIPTION
**Pull request by madhuracj**
Signed-off-by: Madhura Jayaratne madhura.cj@gmail.com

### CODE REVIEW

**Comment by lem9** [comment]
Started testing. I know that I can use the "Find out why" link to create the missing table, but I should not have an error when clicking on Export if I forgot to create it.

**Comment by lem9** [comment]
So, the template is per table? Why?

**Comment by madhuracj** [comment]
@lem9 
Thanks for testing Marc.
On another server where neither the configuration table nor the entry in config.inc.php is setup, I cannot seem to see any errors. Is it on server level export page you get the error?

Templates can be for server (both `db_name` and `table_name` is `NULL`), db (`table_name` is `NULL`) or table.

**Comment by lem9** [comment]
The error happens when exporting at the table level. The generated FROM clause contains only
````
`phpmyadmin`.
````

About templates, when I am at the table level, it would be helpful to load a template that I made for another table. That's the purpose of having a template, right?

**Comment by madhuracj** [comment]
What I had in mind was repeat exports of same database or tables. But I see that ability to use a template made for another db or table can be useful. I'll change accordingly.

**Comment by madhuracj** [comment]
Please have a look now

**Comment by lem9** [comment]
The error is now fixed and the UI behaves correctly, whether I have the missing pmadb or not. I tried table templates and now I can reuse a template for another table. I'll continue testing.

One confusing thing: when I click on Export, I see the Existing templates interface but I have no way of telling whether or not a template has been loaded. There is no difference in the interface, before or after loading (apart from checking if the export options have changed). Maybe the UI should tell me that there is a loaded template (and which one).

**Comment by madhuracj** [comment]
Please check https://github.com/madhuracj/phpmyadmin/commit/f2cf3e5a72d80d54e84e76acd22b5412bc779379 and https://github.com/madhuracj/phpmyadmin/commit/5dbc0c45056f0454773a8426f3d3b6cc5fcc8dcb to see whether the UI has improved

File: libraries/config.default.php [hunk 1]
```diff
@@ -491,6 +491,15 @@
 $cfg['Servers'][$i]['designer_settings'] = '';
 
 /**
+ * table to store export templates
+ *   - leave blank to disable saved searches feature
+ *     SUGGESTED: 'pma__exporttemplates'
+ *
+ * @global string $cfg['Servers'][$i]['exporttemplates']
+ */
+$cfg['Servers'][$i]['exporttemplates'] = '';
```
Comment (ibennetch) [inline libraries/config.default.php]: We're not very consistent with naming these variables regarding whether or not to use an underscore between words; I've added this to https://wiki.phpmyadmin.net/pma/2015-07_Meeting#Naming_style_for_Configuration_Storage_variables . There are more with an underscore than without, so perhaps `export_templates` would be more consistent, but that's just my suggestion.

**Comment by ibennetch** [comment]
I like this feature and have a few suggestions.

* It would be nice to be able to reset to the defaults; say for instance I load a template and realize I don't like it and instead want the defaults; could such a button be added easily?
* If I select to "View output as text" and proceed to the next page by clicking Go, everything happens like I expect. If I then press Back, I'm returned to the Export page, again as expected. All the settings are the same, because phpMyAdmin stores and resets the export options based on what I was just viewing, but the Template dropdown returns to "-- Select a template --"; it would be nice if instead the current template was re-selected.

About Marc's suggestion about the UI showing when a template was loaded, I think your improvement was good but what about going further and also showing some text like "Current template: foo" perhaps under the dropdown? I suppose we'd then have to either hide it or add another text if the export settings are changed from what the template is...so I'm not sure about my idea.

**Comment by ibennetch** [comment]
It's also a bit confusing to save a template in the table export page and not see it listed from the database export page. I understand the reason, and I don't have any suggestion on how to fix it, but as a user it would annoy me.

**Comment by lem9** [comment]
What does "Template wss updated" means?

File: libraries/config.default.php (hunk 1 above)
Comment (madhuracj) [inline libraries/config.default.php]: Thanks for the suggestion. I'll update the table name if required after the IRC meeting.

**Comment by madhuracj** [comment]
@lem9 and @ibennetch 
Thanks for the suggestions, I am attending to them.

**Comment by madhuracj** [comment]
@ibennetch 
Easiest way to load the defaults would be to click on the 'Export' menu link.
Chosen template is correctly selected on 'Back' button now.
About having to separate template to 'server', 'db' and 'table' levels and list only relevant templates, I do not see a good solution either.

**Comment by madhuracj** [comment]
@lem9 
Typo fixed.
Template dropdown auto submits now to load the template.

**Comment by lem9** [comment]
Looks fine, thanks.

**Pull request merged**